real cost for Nevlas with PI

rzulian opened the pull request.

coyotte508 commented on src/reward.ts:
```diff
@@ -69,6 +69,11 @@ export default class Reward {
     return rewards.map(reward => new Reward(-reward.count, reward.type));
   }
 
+  static discount(rewards: Reward[], multiplier: number, type: Resource): Reward[] {
```
Should add a comment, like:

`/** For each reward, if it is of corresponding type, reduce the cost by dividing it by multiplier */`

I would also switch the order of the type / multiplier arguments.

rzulian replied:
@coyotte508 : are you using ```tokenModifier``` in the expansion? If not, I would remove it as it's used only for Nevlas with PI

coyotte508 replied:
Nope, not using it. For now it could be replaced with faction === Nevlas ? 2 : 1.

Later we can make it more flexible if need be.

coyotte508 replied:
I need this PR for the range extension when moving spaceships in the expansion (it uses power)

rzulian commented:
@coyotte508 : I've removed ```tokenModifier``` and added a ```discounted``` property in the boardActions. I've left the ```discount``` method in the ```reward```.
The real problem I see is that this is not backward compatible.

coyotte508 commented:
Yea. Also it means that everywhere that power can be spent, code needs to be added specifically for nevlas (e.g. new board actions, ...) as `spendablePowerTokens()` doesn't take into account the discount. This is a small technical debt for the future. Taklons don't have that issue. 

I think the main issue was the visualisation, and that's why the issue was opened in gaia-viewer at first. Maybe the new player info design will make it a bit less confusing, since there's no 'pw' text in it but three drawn bowls? 

Actually, only now I think that a much simpler solution would be to change ` 0/2/2 pw` to something like` 0/2/2(=4) pw` in the viewer for nevlas.

I'm really sorry about realizing this now and dragging this out until the PR is ready.

rzulian commented:
The only missing visualisation in viewer is actually only the hint in the boardActions, which has to be different when Nevlas with PI.
Managing th legacy games is a different issue. we can either do conditional code, or repairs old games (which I had to do for tests).
On the conditional code we could ad a ```EngineVersion``` property on the game. What do you think?

coyotte508 commented on src/actions.ts:
```diff
@@ -34,23 +34,32 @@ export const freeActionsNevlas = [
+  [BoardAction.Power5]: { cost: "4pw", discounted: "2pw", income: ["2k"] },
+  [BoardAction.Power6]: { cost: "3pw", discounted: "2pw", income: ["1step"] },
+  [BoardAction.Power7]: { cost: "3pw", discounted: "2pw", income: ["2t"] },
+  [BoardAction.Qic1]: { cost: "4q", discounted: "4q", income: ["tech"] },
```
What happens if there's a faction that discounts qic usage (eg -1qic) for board actions?

coyotte508 commented on src/actions.ts:
```diff
@@ -34,23 +34,32 @@ export const freeActionsNevlas = [
-  [BoardAction.Qic1]: { cost: "4q", income: ["tech"] },
-  [BoardAction.Qic2]: { cost: "3q", income: [">fed"] },
-  [BoardAction.Qic3]: { cost: "2q", income: ["3vp", "pt > vp"]}
+  [BoardAction.Power1]: { cost: "7pw", discounted: "4pw", income: ["3k"] },
```
I said that nevlas were the only faction to discount power even in the expansion, but that was assuming a call to something like `discount(rewards, Resource.Power, nevlasDiscount ? 2 : 1)` that could be made more flexible in the future. Not hardcoding the discounted price with a fixed multiplier.

rzulian replied:
Yes, probably your suggestion is better

coyotte508 commented on src/faction-boards/nevlas.ts:
```diff
@@ -15,10 +15,10 @@ const nevlas: FactionBoardRaw = {
     freeActionChoice:  (player: Player, pool: any[]) => {
       pool.push(...freeActionsNevlas);
       if (player.data.hasPlanetaryInstitute()) {
+        pool.length = 0;
```
Nice. I didn't know arrays could be resized this way!

rzulian replied:
Actually, that was the only way to let it work. Everything else wasn't resetting ```pool``` in the emit event :)

coyotte508 replied:
Probably `pool.splice(0)` would have worked. But your way is nicer :)

coyotte508 commented on src/reward.ts:
```diff
@@ -69,6 +69,12 @@ export default class Reward {
   }
 
+  /** For each reward, if it is of corresponding type, reduce the cost by dividing it by multiplier */
+  static discount(rewards: Reward[], type: Resource, multiplier: number): Reward[] {
```
Sadly it seems this function is now unused

rzulian replied:
Unless we use your suggestion above

rzulian commented:
We can put this on hold until we decide what to do. 
The problem I see is that our logs are wrong. We will see.
<img width="1417" alt="screen shot 2018-10-14 at 18 43 41" src="https://user-images.githubusercontent.com/29941924/46949142-ace6d380-d080-11e8-8cdc-46764681fbc8.png">
